### PR TITLE
Proof of concept using Ruby

### PR DESCRIPTION
**Pull request by sergiobayona**
Working example using Ruby to extract unstructured data from a prompt then correctly parsing and loading the response values into a Ruby object. 

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 6471e729d16041eb444297ce055daee8cb5ba5a5.  | 
|--------|--------|

### Summary:
This PR introduces a Ruby implementation for extracting structured data from unstructured text using OpenAI's API, along with usage instructions and basic tests.

**Key points**:
- Added a new `Instructor` module in `/lib/instructor.rb`.
- Introduced `Instructor::Model` class in `/lib/instructor/model.rb` for schema validation.
- Added `Instructor::OpenAI::Client` class in `/lib/instructor/openai/client.rb` for interacting with OpenAI's API.
- Updated `/README.md` with instructions on how to use the new Ruby implementation.
- Included a basic spec in `/spec/basic_spec.rb` for testing the functionality.
----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,49 +17,49 @@ Dive into the world of Ruby-based structured extraction, by OpenAI's function ca
...
+user = client.chat(
+  parameters: {
+    model: 'gpt-3.5-turbo',
+    messages: [{ role: 'user', content: 'Extract Jason is 25 years old' }]
+  },
+  response_model: UserDetail
+)
+
+puts(user.inspect)
+#=> #<Dry::Validation::Result{:name=>"Jason", :age=>25} errors={}>
```
Comment (sergiobayona) [inline README.md:56]: Note the response object is a `Dry::Validate::Result`. We'd probably want an instance of the UserDetail class instead.

File: lib/instructor/openai/client.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+module Instructor
+  module OpenAI
+    class Client
+      def initialize
+        @client = ::OpenAI::Client.new do |f|
+          f.response :logger, Logger.new($stdout), bodies: true if ENV['OPENAI_LOG'] == 'debug'
+        end
+      end
+
+      def chat(parameters:, response_model:)
```
Comment (jxnl) [inline lib/instructor/openai/client.rb:14]: its not `chat.completions` ?

File: lib/instructor/openai/client.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        func = generate_function(response_model)
+        params = parameters.merge(tools: [func])
```
Comment (jxnl) [inline lib/instructor/openai/client.rb:16]: one of the features is the need to have: https://jxnl.github.io/instructor/concepts/patching/

File: lib/instructor/openai/client.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+          function: {
+            name: model.name.humanize.titleize,
+            description: generate_description(model),
+            parameters: model.schema.json_schema
+          }
+        }
+      end
+
+      def get_parsed_res(response)
+        str = response.dig('choices', 0, 'message', 'tool_calls', 0, 'function', 'arguments')
```
Comment (jxnl) [inline lib/instructor/openai/client.rb:35]: https://jxnl.github.io/instructor/concepts/patching/ again, should be matched against the 'mode'

File: lib/instructor/openai/client.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        model = response_model.new
+        model.call(function_response)
+      end
+
+      def generate_function(model)
+        {
+          type: 'function',
+          function: {
+            name: model.name.humanize.titleize,
+            description: generate_description(model),
```
Comment (jxnl) [inline lib/instructor/openai/client.rb:28]: can this be overriden?

File: lib/instructor/openai/client.rb (hunk 2 above)
Comment (sergiobayona) [reply lib/instructor/openai/client.rb:14]: The OpenAI ruby client lib just uses `chat`. For example:

```Ruby
client = OpenAI::Client.new
response = client.chat(
    parameters: {
        model: "gpt-3.5-turbo", 
        messages: [{ role: "user", content: "Hello!"}], 
    })
puts response.dig("choices", 0, "message", "content")
# => "Hello! How may I assist you today?"
    })
```
So I just followed the same pattern. I know the python OpenAI does it differently.

Comment (jxnl) [reply lib/instructor/openai/client.rb:14]: Perfect. That's all I wanted to know.

**Pull request closed without merging**